src/machine/parsed_results.rs

```diff
@@ -141,6 +79,212 @@ pub enum Value {
+        }
+
+        disj_term
+    }
+}
+
+impl From<&QueryResolutionLine> for Value {
+    fn from(qrl: &QueryResolutionLine) -> Self {
+        match qrl {
+            QueryResolutionLine::True => Value::Atom("true".into()),
+            QueryResolutionLine::False => Value::Atom("false".into()),
+            QueryResolutionLine::Match(m) => Value::conjunction(
+                &m.bindings
+                    .iter()
+                    .map(|(k, v)| {
+                        Value::Structure("=".into(), vec![Value::Var(k.clone()), v.clone()])
+                    })
+                    .collect::<Vec<_>>(),
+            ),
+        }
+    }
+}
+
+impl From<&QueryResolution> for Value {
+    fn from(qr: &QueryResolution) -> Self {
+        match qr {
+            QueryResolution::True => Value::Atom("true".into()),
+            QueryResolution::False => Value::Atom("false".into()),
+            QueryResolution::Matches(matches) => Value::disjunction(
+                &matches
+                    .iter()
+                    .map(|m| {
+                        Value::conjunction(
+                            &m.bindings
+                                .iter()
+                                .map(|(k, v)| {
+                                    Value::Structure(
+                                        "=".into(),
+                                        vec![Value::Var(k.clone()), v.clone()],
+                                    )
+                                })
+                                .collect::<Vec<_>>(),
+                        )
+                    })
+                    .collect::<Vec<_>>(),
+            ),
+        }
+    }
+}
+
+struct ValueRationalInner {
+    numerator: IBig,
+    denominator: UBig,
+}
+
+impl Serialize for ValueRationalInner {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        let mut seq = serializer.serialize_seq(Some(2))?;
+        match self.numerator.to_i64() {
+            Some(small) => seq.serialize_element(&small)?,
+            None => seq.serialize_element(&self.numerator.to_string())?,
+        }
+        match self.denominator.to_u64() {
+            Some(small) => seq.serialize_element(&small)?,
+            None => seq.serialize_element(&self.denominator.to_string())?,
+        }
+        seq.end()
+    }
+}
+
+impl Serialize for Value {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        match self {
+            Value::Integer(i) => match i.to_i64() {
+                Some(small) => serializer.serialize_i64(small),
+                None => {
+                    let mut map = serializer.serialize_map(Some(1))?;
+                    map.serialize_entry("integer", &i.to_string())?;
+                    map.end()
+                }
+            },
+            Value::Rational(r) => {
+                let (up, down) = r.clone().into_parts();
```

Comment:
I would have probably also renamed the local variables `up`/`down` here, but this isn't more than just a nit.